src/core/toga/widgets/canvas.py

```diff
@@ -7,11 +7,34 @@
 from ..font import Font, SYSTEM
 
 
-class CanvasContextMixin:
+class Context:
+    """The user-created :class:`Context <Context>` drawing object to populate a
+    drawing with visual context.
+
+    The top left corner of the canvas must be painted at the origin of the
+    context and is sized using the rehint() method.
+
+    """
+
     def __init__(self, *args, **kwargs):  # kwargs used to support multiple inheritance
         super().__init__(*args, **kwargs)
-
         self._canvas = None
+        self.drawing_objects = []
+
+    def __repr__(self):
+        return "{}()".format(self.__class__.__name__)
+
+    def draw(self, impl, *args, **kwargs):
```

Comment:
Should this be `_draw`, rather than `draw`? It isn't a public API that you should be calling.

Reply:
Which linter is giving you the error? Does it give you any reference for the "bug" it found? Because that seems like an error in the linter, rather than a bug in the code. PEP8 describes the single underscore prefix as a 'weak "internal use" indicator'; `import *` won't import a symbol that starts with underscore, but the name is otherwise preserved and available. Using an "internal" symbol across modules should be perfectly acceptable usage, AIUI.

I agree that there are two interpretations of "public" here, but it's the first one that you mention that is most important from my perspective. We're in charge of the internals (and the internal usage), so we can expect a certain amount of cooperation and compliance in the second case; but the first case is where we put things in the hands of the user base at large.

I'm not a huge fan of the Requests approach to API safety. While documentation is an important part of the picture, a big feature of Python as a language is discoverability. I can drop into a prompt at any point, run `dir()`, and find out what is there for me to poke. Seeing `_draw()` implies there's a method that *can* be invoked, but it's probably not intended for me to use. Seeing `draw()` implies there's a method that I can invoke; if the only way to verify that I *shouldn't* be invoking it is to check if it *isn't* mentioned in a piece of documentation that isn't related to the code I'm currently looking at... that's not a very discoverable control.

In this case, `draw()` is very clearly an internal method. It's entirely appropriate for the impl layer to invoke it (after all, that's what the method exists for), but end users shouldn't even know it exists. We can't *completely* hide it, but we can make it obvious from naming that they shouldn't be touching it.

`canvas` and `add_drawing_object()`, on the other hand, aren't really *intended* for public consumption, but it doesn't really matter if the user *does* access them. There in the category of "well, you probably *shouldn't* be poking here, but it doesn't really matter if you do".

`canvas` will be an interface level object, so accessing it isn't a big problem. Modifying it might lead to some entertaining results, but I think they'd at least be *predictable* entertaining results (generally objects disappearing from one canvas, appearing on another).

As for `add_drawing_object()` - while I agree end users generally shouldn't *need* to invoke this one, it shouldn't be inherently dangerous to invoke it. I'd even suggest renaming the method as `add()`. While we don't necessarily want to *encourage* it's use, but it's not *harmful* to put another drawing object onto a context, as long as that object adheres to the interface requirements of a drawing object.

Reply:
Correct - that's my reading of the general intent of PEP8; and if that isn't the intent... that's what I'm running with in Toga's API design :-P